include/seastar/core/abort_source.hh

```diff
@@ -112,8 +112,10 @@ public:
 
     /// Requests that the target operation be aborted. Current subscriptions
     /// are invoked inline with this call, and no new ones can be registered.
+    /// Must be called only once, otherwise terminate will be called.
     void request_abort() noexcept {
-        std::exchange(_subscriptions, std::nullopt)->clear_and_dispose([] (subscription* s) { s->on_abort(); });
+        auto subs = std::exchange(_subscriptions, std::nullopt).value();
```

Comment:
I'm not a fan of all the hidden wisdom that this code contains but won't be shared by 90% of the readers. To understand this code, one has to know that `value()` is special in that unlike `operator*` it checks if there is a value, and if not throws (see https://en.cppreference.com/w/cpp/utility/optional/value). Then you need to understand is that the "noexcept" doesn't mean you think this function won't ever throw - it means you want to terminate if it does, and that it goes with the `value()` call which **can** throw.

I think everything will be much clearer if you just be more explicit on what you want to test. I.e., add `assert(s.has_value())` or simply `assert(s)` at the top (or instead of assert, use internal error or something). Another benefit of using an assert instead of "noexcept" is that you'll get a textual message telling you that the crash happened on request_abort() being double-called, even if you don't get a working backtrace. Isn't this better than a generic "terminate" message?